common/encoding.cpp

```diff
@@ -0,0 +1,339 @@
+#ifdef USE_ICONV
+	if (iconvHandle != (iconv_t) -1)
+		result = convertIconv(iconvHandle, string, length);
+#endif // USE_ICONV
+	if (result == nullptr)
+		result = g_system->convertEncoding(addUtfEndianness(to).c_str(),
+				addUtfEndianness(from).c_str(), string, length);
+
+	if (result == nullptr) {
+		result = convertTransManMapping(addUtfEndianness(to).c_str(), addUtfEndianness(from).c_str(), string, length);
+	}
+
+	return result;
+}
+
+char *Encoding::convertIconv(iconv_t iconvHandle, const char *string, size_t length) {
+#ifdef USE_ICONV
+
+	size_t inSize = length;
+	size_t outSize = inSize;
+	size_t stringSize = inSize > 4 ? inSize : 4;
+
+
+#ifdef ICONV_USES_CONST
+	const char *src = string;
+#else
+	char *src = new char[length];
+	char *originalSrc = src;
+	memcpy(src, string, length);
+#endif // ICONV_USES_CONST
+
+	char *buffer = (char *) calloc(sizeof(char), stringSize);
+	if (!buffer) {
+		warning ("Cannot allocate memory for converting string");
+		return nullptr;
+	}
+	char *dst = buffer;
+	bool error = false;
+
+	while (inSize > 0) {
+		if (iconv(iconvHandle, &src, &inSize, &dst, &outSize) == ((size_t)-1)) {
+			// from SDLs implementation of SDL_iconv_string (slightly altered)
+			if (errno == E2BIG) {
+				char *oldString = buffer;
+				stringSize *= 2;
+				buffer = (char *) realloc(buffer, stringSize);
+				if (!buffer) {
+					warning ("Cannot allocate memory for converting string");
+					error = true;
+					break;
+				}
+				dst = buffer + (dst - oldString);
+				outSize = stringSize - (dst - buffer);
+				memset(dst, 0, stringSize / 2);
+			} else {
+				error = true;
+				break;
+			}
+		}
+	}
+	iconv(iconvHandle, NULL, NULL, &dst, &outSize);
+	// Add a zero character to the end. Hopefuly UTF32 uses the most bytes from
+	// all possible encodings, so add 4 zero bytes.
+	buffer = (char *) realloc(buffer, stringSize + 4);
+	memset(buffer + stringSize, 0, 4);
+
+#ifndef ICONV_USES_CONST
+	delete[] originalSrc;
+#endif // ICONV_USES_CONST
+
+	if (error) {
+		if (buffer)
+			free(buffer);
+		return nullptr;
+	}
+	return buffer;
+#else
+	return nullptr;
+#endif //USE_ICONV
+}
+
+// This algorithm is able to convert only between the current TransMan charset
+// and UTF-32, but if it fails, it tries to at least convert from the current
+// TransMan encoding to UTF-32 and then it calls convert() again with that.
+char *Encoding::convertTransManMapping(const char *to, const char *from, const char *string, size_t length) {
+#ifdef USE_TRANSLATION
+	String currentCharset = TransMan.getCurrentCharset();
+	if (currentCharset.equalsIgnoreCase(from)) {
+		// We can use the transMan mapping directly
+		uint32 *partialResult = (uint32 *) calloc(sizeof(uint32), (strlen(string) + 1));
+		if (!partialResult) {
+			warning("Couldn't allocate memory for encoding conversion");
+			return nullptr;
+		}
+		const uint32 *mapping = TransMan.getCharsetMapping();
+		if (mapping == 0) {
+			for(unsigned i = 0; i < strlen(string); i++) {
```

Comment:
Using strlen in loop conditional is suboptimal as string gets looped over at each for loop iteration.
You should store the result of strlen before looping.